<div style="direction:rtl">

<p> در این پروژه ، دو نوع مختلف از پست های وبسایت Hacker News (که  در حوزه تکنولوژی معروف است) را با یکدیگر مقایسه خواهیم کرد. دو نوع پست مورد بررسی با Ask HN یا Show HN شروع می شوند. </p> 

<p> کاربران با ارسال پست های HN Ask جامعه Hacker News سوال می پرسند ، مانند "بهترین دوره آنلاین که تاکنون گذرانده اید چیست؟" . به همین ترتیب ، کاربران با استفاده از پست های HN Show به جامعه Hacker News یک پروژه ، محصول یا موارد جالب توجه را نشان می دهند. </p>

<p> ما برای تعیین موارد زیر به طور خاص این دو نوع پست را با یکدیگر مقایسه خواهیم کرد: </p>

<p>     آیا Ask HN یا Show HN به طور متوسط کامنت های بیشتری دریافت می کنند؟
    آیا پست های ایجاد شده در یک زمان خاص به طور متوسط نظرات بیشتری دریافت کرده اند؟ </p>

<p> لازم به ذکر است که مجموعه داده هایی که ما با آن کار می کنیم با حذف مواردی که هیچ نظری دریافت نکرده اند و سپس به صورت تصادفی از ارسال های باقیمانده ، از تقریباً 300000 ردیف به تقریباً 20000 ردیف کاهش یافته است. </p>
</div>

# مقدمه


In [3]:
# Read in the data.
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

# حذف عناوین از لیست تودرتو

In [4]:
# Remove the headers.
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


<div style="direction:rtl">

در بالا می بینیم که دیتاست حاوی عناوین پست ها ، تعداد نظرات برای هر پست و تاریخ ایجاد پست است. بیایید با بررسی تعداد نظرات برای هر نوع پست شروع کنیم.
    </div>

# استخراج پست های پرسشی و نمایشی


<div style="direction:rtl">
ابتدا پست هایی را که با "Ask HN" یا "Show HN" شروع می شوند شناسایی می کنیم و داده های مربوط به این دو نوع پست را در لیست های مجزا قرار می دهیم. تفکیک داده ها ، انجام مراحل زیر را آسان تر می کند.
</div>



In [5]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts =[]
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


# محاسبه میانگین تعداد کامنت پست های پرسشی و نمایشی

<div style="direction:rtl">
اکنون که پست های پرسشی و نمایشی را در لیست های مجزا قرار دادیم، میانگین تعداد نظرات هر دو نوع پست را محاسبه خواهیم کرد.
</div>


In [6]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [7]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


<div style="direction:rtl">
به طور متوسط ، پست های پرسشی در نمونه ما تقریباً 14 کامنت دریافت می کنند ، در حالی که پست های نمایشی تقریباً 10 کامنت دریافت می کنند. از آنجا که پست های پرسشی بیشتر احتمال دارد کامنت دریافت کنند ، ادامه تجزیه و تحلیل را فقط بر روی این نوع پست ها متمرکز می کنیم.</div>


# یافتن تعداد پست های پرسشی و کامنت ها بر اساس ساعت


<div style="direction:rtl">
حال می خواهیم ببینیم که آیا می توانیم با ایجاد پست در یک زمان خاص ، مقداربیشتری  کامنت دریافت کنیم. در ابتدا ، میزان پست های پرسشی که در هر ساعت از شبانه روز ایجاد شده اند را به همراه تعداد کامنت هایی که این پست ها دریافت کرده اند ، خواهیم یافت. سپس ، میانگین مقدار کامنت های ارسال شده در هر ساعت از شبانه روز را محاسبه خواهیم کرد.
</div>


In [8]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append(
        [post[6], int(post[4])]
    )

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    date = each_row[0]
    comment = each_row[1]
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    if time in counts_by_hour:
        comments_by_hour[time] += comment
        counts_by_hour[time] += 1
    else:
        comments_by_hour[time] = comment
        counts_by_hour[time] = 1

comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

# محاسبه میانگین تعداد کامنت های پست های پرسشی بر اساس ساعت

In [9]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

# مرتب کردن و پرینت مقادیر از یک لیست تودرتو

In [10]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [11]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


<div style="direction:rtl">
ساعتی که به طور متوسط بیشترین کامنت را به ازا هر پست دریافت می کند ، ساعت 15:00 است که به طور متوسط 38.59 نظر در هر پست دارد. ساعت دارای رتبه اول %60 از رتبه دوم تعداد کامنت بیشتری دریافت کرده است. 
</div>


# نتیجه گیری



<div style="direction:rtl">
در این پروژه ، ما پست های پرسشی و نمایشی را تجزیه و تحلیل کردیم تا مشخص کنیم کدام نوع پست و کدام  زمان، بیشترین کامنت را دریافت می کند. بر اساس تجزیه و تحلیل ما ، برای به حداکثر رساندن شانس دریافت نظرات ، توصیه می کنیم پست به عنوان پست سوالی طبقه بندی شود و بین ساعت 15:00 تا 16:00 ایجاد شود (3:00 بعد از ظهر - 4:00 بعد از ظهر).

با این حال ، لازم به ذکر است که در دیتاستی که مورد تجزیه و تحلیل قرار دادیم، پست های بدون کامنت حذف شده  اند. با توجه به این ، گفتن جمله "از پستهایی که نظرات دریافت کرده اند"  دقیق تر است.  پست های پرسشی به طور متوسط کامنت بیشتری دریافت کرده اند و پستهای ایجاد شده بین ساعت 15:00 تا 16:00 به طور متوسط بیشترین کامنت را دریافت کرده اند.
</div>
